src/components/CardGrid.tsx

```diff
@@ -328,53 +324,54 @@ export function CardGrid<T>(props: CardGridProps<T>): JSX.Element {
       break;
     }
   }
+  const data = props.dataContext.data || [];
 
-  // set the body of the grid based on data loading state
-  useEffect(() => {
-    if (props.dataContext.loading) {
-      // loading previews
-      setBodyElements(
-        Array(gridCols)
-          .fill(null)
-          .map((_) => props.cardContext.loadingCardCreator())
-      );
-    } else if (props.dataContext.data === undefined || props.dataContext.data.length === 0) {
-      // no-data fallback
-      setBodyElements([props.cardContext.noDataFallback ?? <></>]);
-    } else {
-      // loaded data
-      setBodyElements(
-        props.dataContext.data.map((cardData) =>
-          props.cardContext.cardCreator(
-            cardData,
-            props.dataContext.refetch,
-            props.dataContext.loading === undefined ? true : props.dataContext.loading
-          )
-        )
-      );
-    }
-  }, [setBodyElements, props, gridCols, gridId]);
+  if (!props.dataContext.loading && data.length === 0) {
+    return (
+      <div className="w-full pb-8">
+        <div className="flex w-full justify-center p-8">{props.cardContext.noDataFallback}</div>
+      </div>
+    );
+  }
 
   return (
-    <div className="w-full justify-center">
+    <div className="w-full pb-8">
       <div
         className={clsx('grid grid-cols-1 gap-6 sm:grid-cols-2', gridViewClasses, props.className)}
       >
-        {bodyElements.map((e, i) => (
-          <div key={`${gridId}-${i}`}>{e}</div>
+        {data.map((e, i) => (
+          <div key={`${gridId}-${i}`}>
+            {props.cardContext.cardCreator(e, props.dataContext.refetch, false)}
+          </div>
         ))}
-      </div>
-      <InView threshold={0.1} onChange={props.dataContext.onLoadMore} className="w-full">
-        <div
-          className={clsx('my-6 flex w-full items-center justify-center font-bold', {
-            hidden:
-              !props.dataContext.hasMore ||
-              (props.dataContext.data && props.dataContext.data.length === 0),
+        {props.dataContext.loading &&
+          [...Array(gridCols * gridCols).keys()].map((i) => {
```

Comment:
```suggestion
          [...Array(gridCols * gridRows).keys()].map((i) => {
```

Reply:
There is no grid rows. Are you suggesting that is added?

Reply:
Ah, no. I was thinking of this as a fixed-size grid when I saw that